src/Translator/assets/src/formatters/formatter.ts

```diff
@@ -0,0 +1,238 @@
+        }
+    }
+
+    const intervalRegex = /^(?<interval>({\s*(-?\d+(\.\d+)?[\s*,\s*\-?\d+(.\d+)?]*)\s*})|(?<left_delimiter>[[\]])\s*(?<left>-Inf|-?\d+(\.\d+)?)\s*,\s*(?<right>\+?Inf|-?\d+(\.\d+)?)\s*(?<right_delimiter>[[\]]))\s*(?<message>.*?)$/s;
+
+    const standardRules: Array<string> = [];
+    for (let part of parts) {
+        part = part.trim().replace(/\|\|/g, '|');
+
+        let matches = part.match(intervalRegex);
+        if (matches !== null) {
+            if (matches[2]) {
+                for (const n of matches[3].split(',')) {
+                    if (number === Number(n)) {
+                        return strtr(matches.groups!['message'], parameters);
+                    }
+                }
+            } else {
+                const leftNumber = '-Inf' === matches.groups!['left'] ? Number.NEGATIVE_INFINITY : Number(matches.groups!['left']);
+                const rightNumber = ['Inf', '+Inf'].includes(matches.groups!['right']) ? Number.POSITIVE_INFINITY : Number(matches.groups!['right']);
+
+                if (('[' === matches.groups!['left_delimiter'] ? number >= leftNumber : number > leftNumber)
+                    && (']' === matches.groups!['right_delimiter'] ? number <= rightNumber : number < rightNumber)
+                ) {
+                    return strtr(matches.groups!['message'], parameters);
+                }
+            }
+        } else {
+            matches = part.match(/^\w+:\s*(.*?)$/);
+            if (matches !== null) {
+                standardRules.push(matches[1]);
+            } else {
+                standardRules.push(part);
+            }
+        }
+    }
+
+    const position = getPluralizationRule(number, locale);
+    if (typeof standardRules[position] === 'undefined') {
+        // when there's exactly one rule given, and that rule is a standard
+        // rule, use this rule
+        if (1 === parts.length && typeof standardRules[0] !== 'undefined') {
+            return strtr(standardRules[0], parameters);
+        }
+
+        throw new Error(`Unable to choose a translation for "${id}" with locale "${locale}" for value "${number}". Double check that this translation has the correct plural options (e.g. "There is one apple|There are %count% apples").`)
+    }
+
+    return strtr(standardRules[position], parameters);
+}
+
+/**
+ * Returns the plural position to use for the given locale and number.
+ *
+ * The plural rules are derived from code of the Zend Framework (2010-09-25),
+ * which is subject to the new BSD license (http://framework.zend.com/license/new-bsd).
+ * Copyright (c) 2005-2010 Zend Technologies USA Inc. (http://www.zend.com)
+ */
+function getPluralizationRule(number: number, locale: Locale): number {
+    number = Math.abs(number);
+    let _locale = locale;
+
+    if (locale === 'pt_BR' || locale === 'en_US_POSIX') {
+        return 0;
+    }
+
+    _locale = _locale.length > 3 ? _locale.substring(0, _locale.indexOf('_')) : _locale;
+
+    switch (_locale) {
```

Comment:
Is there no way to extract this switch in a specific function to avoid duplicating it both here and in `translator_controller` ? Easier to maintain and unit-test.

Reply:
The `formatter.ts` is the only file which contains this logic, the `translator_controller` you are reffering the the one from `dist/` folder, which basically contains all the code from `src/` folder.